src/core/src/package_managers/AptitudePackageManager.py

```diff
@@ -79,9 +80,55 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
-        self.invoke_package_manager(self.repo_refresh)
+    # region Sources Management
+    def __get_custom_sources_to_spec(self, max_patch_published_date=str(), base_classification=str()):
+        # type: (str, str) -> str
+        """ Prepares the custom sources list for use in a command. Idempotent. """
+        try:
+            formula = "F-[{0}]-[{1}]".format(max_patch_published_date, base_classification)
+            if self.cached_customer_source_list_formula == formula:
+                return self.custom_sources_list
+            self.cached_customer_source_list_formula = formula
+
+            command = self.cmd_prep_custom_sources_list_template.replace("<CLASSIFICATION>", base_classification if base_classification == "security" else "\"\"")
+            code, out = self.env_layer.run_command_output(command, False, False)
+            sources_content = self.env_layer.file_system.read_with_retry(self.custom_sources_list)
+            self.composite_logger.log_debug("[APM] Modified custom sources list with classification. [Code={0}][Out={1}][Content={2}]".format(str(code), str(out), str(sources_content)))   # non-zero error code to be investigated
+
+            if max_patch_published_date != str():
+                target = "://snapshot.ubuntu.com/ubuntu/{0}".format(self.max_patch_publish_date)
+
+                if "://snapshot.ubuntu.com/ubuntu/" in sources_content:
+                    sources_content_split = sources_content.split(" ")
+                    for i in range(0, len(sources_content_split)):
+                        if "://snapshot.ubuntu.com/ubuntu/" in sources_content_split[i]:
+                            sources_content.replace(sources_content_split[i], target)
+
+                sources_content = sources_content.replace("://azure.archive.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("://in.archive.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("://security.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("http://snapshot.ubuntu.com/", "https://snapshot.ubuntu.com/")
+                self.composite_logger.log_debug("[APM] Modified custom sources list with snapshot. [Code={0}][Out={1}][Content={2}]".format(str(code), str(out), str(sources_content)))
+
+                self.env_layer.file_system.write_with_retry_using_temp_file(self.custom_sources_list, sources_content)
+
+            self.refresh_repo(sources=self.custom_sources_list)
+        except Exception as error:
+            self.composite_logger.log_error("[APM] Error in modifying custom sources list. [Error={0}]".format(repr(error)))
+            return str()    # defaults code to safety
+
+        return self.custom_sources_list
+
+    def refresh_repo(self, sources=str()):
+        self.composite_logger.log("[APM] Refreshing local repo... [Sources={0}]".format(sources if sources != "" else "Default"))
```

Comment:
if sources != ""

Should this be if sources != str()?

Using if new_sources_list != str() on line 130

Reply:
str() and "" are the same thing. I just started having a preference to the former. Can be addressed in either direction in the bigger PR. If I need to do an iteration on this one, I'll swap out, or leave as-is in this one.

Reply:
Addressed since new iteration is needed.